### PR TITLE
Point-to-point health check. Client side implementation

### PR DESCRIPTION
**Pull request by MakMukhi**

### CODE REVIEW

File: transport/http2_client.go [hunk 1]
```diff
@@ -98,6 +100,11 @@ type http2Client struct {
 
 	creds []credentials.PerRPCCredentials
 
+	// activity counter
+	activity *uint64
```
Comment (Thomasdezeeuw) [inline transport/http2_client.go]: It's common to just use an `uint64`, rather then a reference. For example see http.Server code: https://github.com/golang/go/blob/7dc97d9e328edc800e2ce41d5b211ef4e0ef41d0/src/net/http/server.go#L2363-L2364.

File: keepalive/keepalive.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+// DefaultKParams contains default values for keepalive parameters
+var DefaultKParams = Params{
+	Ktime:     time.Duration(math.MaxInt64),           // default to infinite
+	Ktimeout:  time.Duration(20 * 1000 * 1000 * 1000), // default to 20 seconds
+	KNoStream: false,
+}
+
+// Mu is a mutex to protect Enabled variable
+var Mu = sync.Mutex{}
```
Comment (Thomasdezeeuw) [inline keepalive/keepalive.go]: Rather then exposing this lock, why not use functions for it. e.g.

```go
func Enabled() bool {}
func Enable() {}
func Disable() {}
```

File: keepalive/keepalive.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+	Ktime time.Duration
+	// After having pinged fot keepalive check, the client waits for a duration of keepalive_timeout before closing the transport.
+	Ktimeout time.Duration
+	//If true, client runs keepalive checks even with no active RPCs.
+	KNoStream bool
+}
+
+// DefaultKParams contains default values for keepalive parameters
+var DefaultKParams = Params{
+	Ktime:     time.Duration(math.MaxInt64),           // default to infinite
```
Comment (MakMukhi) [inline keepalive/keepalive.go]: Setting it to -1 would fire instantly.

File: keepalive/keepalive.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import (
+	"math"
+	"sync"
+	"time"
+)
+
+// Params is used to set keepalive parameters.
+type Params struct {
+	// After a duration of this time the client pings the server to see if the transport is still alive.
+	Time time.Duration
```
Comment (ejona86) [inline keepalive/keepalive.go]: Need to think about what happens when these are 0.

File: keepalive/keepalive.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+	// After having pinged fot keepalive check, the client waits for a duration of keepalive_timeout before closing the transport.
+	Timeout time.Duration
+	//If true, client runs keepalive checks even with no active RPCs.
+	PermitWithoutStream bool
+}
+
+// DefaultParams contains default values for keepalive parameters.
+var DefaultParams = Params{
```
Comment (ejona86) [inline keepalive/keepalive.go]: Consider moving to transport.

File: keepalive/keepalive.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+// mu is a mutex to protect Enabled variable.
+var mu = sync.Mutex{}
+
+// enable is a knob used to turn keepalive on or off.
+var enable = false
+
+// Enabled exposes the value of enable variable.
+func Enabled() bool {
```
Comment (ejona86) [inline keepalive/keepalive.go]: This seems unnecessary. I don't know a case where WithKeepaliveParams wouldn't work.

File: transport/http2_client.go [hunk 7]
```diff
@@ -976,6 +989,7 @@ func (t *http2Client) reader() {
 	// loop to keep reading incoming messages on this transport.
```
Comment (ejona86) [inline transport/http2_client.go]: Update activity after handleSettings?

File: transport/http2_client.go [hunk 8]
```diff
@@ -1082,6 +1105,26 @@ func (t *http2Client) controller() {
 			case <-t.shutdownChan:
 				return
 			}
+		case <-timer.C:
+			t.mu.Lock()
+			ns := len(t.activeStreams)
+			t.mu.Unlock()
+			// Get the activity counter value and reset it.
+			a := atomic.SwapUint64(&t.activity, 0)
+			if a > 0 || (!t.kp.PermitWithoutStream && ns < 1) {
```
Comment (ejona86) [inline transport/http2_client.go]: This doesn't seem to follow this part of the spec:
> Since keepalive is not occurring on HTTP/2 connections without any streams, there will be a higher chance of failure for new RPCs following a long period of inactivity. To reduce the tail latency for these RPCs, it is important to not reset the `keepalive time` when a connection becomes active; if a new stream is created and there has been greater than `keepalive time` since the last read byte, then a keepalive PING should be sent (ideally before the HEADERS frame). Doing so detects the broken connection with a latency of `keepalive timeout` instead of `keepalive time + timeout`.

File: transport/http2_client.go [hunk 9]
```diff
@@ -1070,64 +1091,94 @@ func (t *http2Client) applySettings(ss []http2.Setting) {
...
-						t.applySettings(i.ss)
-					} else {
-						t.framer.writeSettings(true, i.ss...)
+			// If controlMsg is of type fireKeepaliveTimer,
+			// then check if the keepaliveSkipped flag is still set.
+			if _, ok := controlMsg.(fireKeepaliveTimer); ok {
+				if atomic.LoadUint32(&t.keepaliveSkipped) == 1 {
+					// Reset the timer to 0 so that it fires.
+					if !timer.Stop() {
+						<-timer.C
```
Comment (ejona86) [inline transport/http2_client.go]: Could you make a note that this is safe since the timer will always be Reset in `case <-timer.C`?

Otherwise it could deadlock. https://play.golang.org/p/4CHuWZJFwu

File: transport/http2_client.go [hunk 10]
```diff
@@ -1070,64 +1091,94 @@ func (t *http2Client) applySettings(ss []http2.Setting) {
...
+					timer.Reset(0)
 				}
-				t.writableChan <- 0
 				continue
-			case <-t.shutdownChan:
-				return
 			}
+			wchan = t.writableChan
+			cchan = nil
+			continue
```
Comment (ejona86) [inline transport/http2_client.go]: This `continue` has no effect. Ditto in next case. Remove?

File: keepalive/keepalive.go [hunk 11]
```diff
@@ -16,35 +15,20 @@ type Params struct {
...
-	Time:    time.Duration(math.MaxInt64), // default to infinite.
-	Timeout: time.Duration(20 * time.Second),
+// Validate is used to validate the keepalive parameters.
+// Time durations initialized to 0 will be replaced with default Values.
+func (p *Params) Validate() {
+	if p.Time == 0 {
+		p.Time = Infinity
+	}
+	if p.Timeout == 0 {
+		p.Time = TwentySec
```
Comment (ejona86) [inline keepalive/keepalive.go]: s/Time/Timeout/

File: transport/http2_client.go [hunk 12]
```diff
@@ -1070,64 +1091,94 @@ func (t *http2Client) applySettings(ss []http2.Setting) {
 // frames (e.g., window update, reset stream, setting, etc.) to the server.
 func (t *http2Client) controller() {
 	timer := time.NewTimer(t.kp.Time)
-	if !keepalive.Enabled() {
+	if t.kp.Timeout == keepalive.Infinity {
```
Comment (ejona86) [inline transport/http2_client.go]: s/Timeout/Time/

File: transport/http2_client.go [hunk 13]
```diff
@@ -1070,64 +1091,94 @@ func (t *http2Client) applySettings(ss []http2.Setting) {
...
+			wchan <- 0
+			wchan = nil
+			cchan = t.controlBuf.get()
+			continue
 		case <-timer.C:
 			t.mu.Lock()
 			ns := len(t.activeStreams)
 			t.mu.Unlock()
+			if !t.kp.PermitWithoutStream && ns < 1 {
+				timer.Reset(t.kp.Time)
```
Comment (ejona86) [inline transport/http2_client.go]: Hmm... Won't this cause the timer to continue firing on an idle channel? It also makes the various states hard to follow. Could this just be reset to Inifinity?

File: keepalive/keepalive.go [hunk 14]
```diff
@@ -0,0 +1,30 @@
+package keepalive
```
Comment (ejona86) [inline keepalive/keepalive.go]: The separate package looks weird to me as well. We may want to figure out alternatives.

File: transport/http2_client.go [hunk 15]
```diff
@@ -1068,37 +1091,112 @@ func (t *http2Client) applySettings(ss []http2.Setting) {
...
+			// Send ping.
+			t.controlBuf.put(keepalivePing)
+			isPingSent = true
+			timer.Reset(t.kp.Timeout)
 		case <-t.shutdownChan:
+			// stop the keepalive timer
+			if !timer.Stop() {
+				select {
+				case <-timer.C:
+				default:
```
Comment (ejona86) [inline transport/http2_client.go]: So note this prevents this goroutine from blocking permanently, but it may not free the timer. The Stop() may not process immediately, in which case `<-timer.C` doesn't have a value _yet_, but it will in the future.

It's probably better to either get rid of the `if t.kp.Time == infinity {` optimization above or check a second time here (you could have a variable, say `isTimerUsed`).

File: keepalive.go [hunk 16]
```diff
@@ -0,0 +1,15 @@
+package grpc
+
+import (
+	"time"
+)
+
+// KeepaliveParameters is used to set keepalive parameters.
+type KeepaliveParameters struct {
+	// After a duration of this time the client pings the server to see if the transport is still alive.
```
Comment (ejona86) [inline keepalive.go]: It seems there should be some mention of the current defaults. Here and below. Try to use language that implies the defaults may change, like "the current default is" and similar.

**Comment by MakMukhi** [comment]
Thanks for the comments, Eric. On it!

File: transport/http2_client.go (hunk 15 above)
Comment (MakMukhi) [inline transport/http2_client.go]: Since we exit the controller function right after, wouldn't the timer be eventually gc'd. Also, since nowhere outside the function is timer.C accessed, it should be safe even if timer.C later gets a value. In fact, maybe we can get rid of the case <- timer.c: altogether?

Comment (ejona86) [inline transport/http2_client.go]: You're forgetting there is a sender of that value. The sender could keep the timer alive and prevent GC.

Now, I happen to know that the [channel is buffered](https://golang.org/src/time/sleep.go#L75), and so in this specific case the sender will be able to queue an entry and then remove the timer from its data structures and then the timer could be GCed.

File: keepalive/keepalive.go [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+
+import (
+	"time"
+)
+
+// ClientParameters is used to set keepalive parameters for the client side.
+type ClientParameters struct {
+	// After a duration of this time the client pings the server to see if the transport is still alive.
+	Time time.Duration
+	// After having pinged fot keepalive check, the client waits for a duration of keepalive_timeout before closing the transport.
```
Comment (tamird) [inline keepalive/keepalive.go]: s/fot/for/

File: keepalive/keepalive.go [hunk 18]
```diff
@@ -0,0 +1,15 @@
...
+	Timeout time.Duration
+	//If true, client runs keepalive checks even with no active RPCs.
```
Comment (tamird) [inline keepalive/keepalive.go]: missing space

File: keepalive/keepalive.go (hunk 17 above)
Comment (tamird) [inline keepalive/keepalive.go]: would be nice to wrap this.

File: transport/http2_client.go [hunk 19]
```diff
@@ -182,6 +196,14 @@ func newHTTP2Client(ctx context.Context, addr TargetInfo, opts ConnectOptions) (
 	if opts.UserAgent != "" {
 		ua = opts.UserAgent + " " + ua
 	}
+	kp := opts.KeepaliveParams
+	// Validate keepalive parameters.
```
Comment (tamird) [inline transport/http2_client.go]: this is setting defaults, not validating.

File: transport/http2_client.go [hunk 20]
```diff
@@ -80,6 +82,8 @@ type http2Client struct {
 	// goAway is closed to notify the upper layer (i.e., addrConn.transportMonitor)
 	// that the server sent GoAway on this transport.
 	goAway chan struct{}
+	// awakenKeepalive is used to tell keepalive goroutine to reset keepalive timer.
+	awakenKeepalive chan int
```
Comment (tamird) [inline transport/http2_client.go]: why is this int instead of `struct{}`? the value is never used.

File: transport/http2_client.go [hunk 21]
```diff
@@ -1104,6 +1143,63 @@ func (t *http2Client) controller() {
 	}
 }
 
+// keepalive running in a separate goroutune makes sure the connection is alive by sending pings.
+func (t *http2Client) keepalive() {
+	if t.kp.Time == time.Duration(math.MaxInt64) {
```
Comment (tamird) [inline transport/http2_client.go]: this should use `infinity` instead of repeating `time.Duration(math.MaxInt64)`

File: transport/http2_client.go [hunk 22]
```diff
@@ -1104,6 +1138,63 @@ func (t *http2Client) controller() {
...
+		return
+	}
+	p := &ping{data: [8]byte{}}
+	timer := time.NewTimer(t.kp.Time)
```
Comment (tamird) [inline transport/http2_client.go]: `defer timer.Stop()` below this line.

File: transport/http2_client.go [hunk 23]
```diff
@@ -1104,6 +1138,63 @@ func (t *http2Client) controller() {
...
+		case <-timer.C:
+			if a := atomic.SwapUint64(&t.activity, 0); a > 0 {
+				timer.Reset(t.kp.Time)
+				continue
+			}
+			// Check if keepalive should go dormant.
+			t.mu.Lock()
+			if len(t.activeStreams) < 1 && !t.kp.PermitWithoutStream {
+				// Make awakenKeepalive writable.
+				<-t.awakenKeepalive
```
Comment (tamird) [inline transport/http2_client.go]: why does this happen under lock? seems clearer to write:
```
t.mu.Lock()
dormant := len(t.activeStreams) < 1 && !t.kp.PermitWithoutStream
t.mu.Unlock()
if dormant {
   ...
} else {
  t.controlBuf.put(p)
}
```

File: transport/http2_client.go [hunk 24]
```diff
@@ -1104,6 +1138,63 @@ func (t *http2Client) controller() {
...
+					return
+				}
+			} else {
+				t.mu.Unlock()
+				// Send ping.
+				t.controlBuf.put(p)
+			}
+
+			// By the time control gets here a ping has been sent one way or the other.
+			timer.Reset(t.kp.Timeout)
```
Comment (tamird) [inline transport/http2_client.go]: why is there so much duplicated code here?

File: transport/transport.go [hunk 25]
```diff
@@ -606,3 +611,10 @@ func wait(ctx context.Context, done, goAway, closing <-chan struct{}, proceed <-
 		return i, nil
 	}
 }
+
+// keepalive related constants.
+const (
+	infinity                = time.Duration(math.MaxInt64)
+	defaultKeepaliveTime    = infinity
+	defaultKeepaliveTimeout = time.Duration(20 * time.Second)
```
Comment (tamird) [inline transport/transport.go]: `time.Duration` is not necessary here.

File: transport/transport_test.go [hunk 26]
```diff
@@ -262,13 +267,130 @@ func setUp(t *testing.T, port int, maxStreams uint32, ht hType) (*server, Client
...
+		PermitWithoutStream: true,            // Run keepalive even with no RPCs.
+	}}, done)
+	defer tr.Close()
+	conn, ok := <-done
+	if !ok {
+		t.Fatalf("Server didn't return connection object")
+	}
+	defer conn.Close()
+	// Sleep for keepalive to close the connection.
+	time.Sleep(4 * time.Second)
```
Comment (tamird) [inline transport/transport_test.go]: perhaps this should be a function of the settings above, rather than hardcoded.

there are 4 instances of this.

File: transport/http2_client.go (hunk 22 above)
Comment (MakMukhi) [inline transport/http2_client.go]: No all exit paths of this routine have a timer running. Trying to stop and empty timer.C of an already stopped timer leads to a deadlock.
Doing this in defer would mean making sure that all exit paths of the routine reset the timer(if one is not running already). Future updates to the code might miss that and get into trouble.

File: transport/http2_client.go (hunk 23 above)
Comment (MakMukhi) [inline transport/http2_client.go]: Emptying of the awakenKeepalive channel needs to happen inside the lock so that some other thread creating a new stream sees the channel as writable if the number of streams go form 0 to 1.

File: transport/http2_client.go (hunk 24 above)
Comment (MakMukhi) [inline transport/http2_client.go]: Duplication here is by choice to keep the structure of the routine simple so that new users can understand the logic behind easily.
To reuse the outer select we'll need to have some additional book-keeping about if the ping was sent or not.

File: transport/http2_client.go [hunk 27]
```diff
@@ -373,6 +396,15 @@ func (t *http2Client) NewStream(ctx context.Context, callHdr *CallHdr) (_ *Strea
 	s := t.newStream(ctx, callHdr)
 	s.clientStatsCtx = userCtx
 	t.activeStreams[s.id] = s
+	// If the number of active streams change from 0 to 1, then check if keepalive
+	// has gone dormant. If so, wake it up.
+	if len(t.activeStreams) == 1 {
+		select {
+		case t.awakenKeepalive <- struct{}{}:
+			t.framer.writePing(true, false, [8]byte{})
```
Comment (menghanl) [inline transport/http2_client.go]: We don't need to flush the ping, we will flush the headers later.

File: transport/http2_client.go [hunk 28]
```diff
@@ -992,6 +1025,7 @@ func (t *http2Client) reader() {
 	// loop to keep reading incoming messages on this transport.
 	for {
 		frame, err := t.framer.readFrame()
+		atomic.AddUint64(&t.activity, 1)
```
Comment (menghanl) [inline transport/http2_client.go]: CAS
